Source/NameServer/SBase.cpp

```diff
@@ -0,0 +1,1550 @@
+   string devTypName = msg->Zname(2); // device type name is in the third field
+   DTypeIdx devIdx = 0;
+   // search through loaded task info for the device type
+   while ((task.DeviceTypes[devIdx].Name != devTypName) &&
+	  (++devIdx < task.DeviceTypes.size()));
+   if (devIdx >= task.DeviceTypes.size()) // this device type was never loaded
+   {
+      Post(718,taskName,int2str(Urank),devTypName); // so error.
+      return AddressBook::ERR_INVALID_DEVTYPE;
+   }
+   vector<string> devNames;     // names and addresses will be in packed vectors
+   vector<SymAddr_t> devAddrs;
+   int numSupers = 1;           // get method expects a *reference* to the count
+   // supervisor of this device is in the fourth dynamic field
+   SymAddr_t* devSuper = msg->Get<SymAddr_t>(3,numSupers); 
+   if (!devSuper)
+   {
+      // no supervisor found
+      Post(718,taskName,int2str(Urank));
+      return ERR_INVALID_SUPERVISOR;
+   }
+   vector<AttrIdx> devAttrs;
+   msg->GetX(0, devNames); // grab device names and addresses out of the first
+   msg->Get(1, devAddrs);  // and second dynamic fields.
+   if (devNames.size() != devAddrs.size()) // mismatching counts is an error
+   {
+      Post(716,int2str(Urank),uint2str(devNames.size()),uint2str(devAddrs.size()),"addresse");
+      return ERR_DEVICE_DATA_MISMATCH;	   
+   }
+   msg->Get(2, devAttrs); // get the device attributes
+   // for which there should be one for each device
+   if (devNames.size() != devAttrs.size()) 
+   {
+      Post(716,int2str(Urank),uint2str(devNames.size()),uint2str(devAttrs.size()),"attribute");
+      return ERR_DEVICE_DATA_MISMATCH;	   
+   }
+   unsigned err = AddressBook::SUCCESS;
+   // everything checks out. Build device objects.
+   for (unsigned d = 0; d < devNames.size(); d++) 
+   {
+       Record_t device(devNames[d],
+		       devAddrs[d],
+		       *devSuper,
+		       devIdx,
+		       RecordType_t(Device),
+		       devAttrs[d]);
+       // and add them to the database
+       if ((err = AddDevice(taskName, device)) != AddressBook::SUCCESS)
+       {
+	  // try to clean out the task if a device couldn't be inserted 
+	  if (ClearTask(taskName) != AddressBook::SUCCESS)
+	  {
+	     Post(708, device.Name, taskName, int2str(Urank));
+	     return err;
+	  }
+	  Post(707, device.Name, taskName, int2str(Urank));
+	  return err;
+       }
+   }
+   if (TaskState(taskName) < Linked) // adding the devices auto-links the records
+   {
+      if ((err = TaskState(taskName,Linked)))
+      {
+	 Post(724,taskName,int2str(Urank));
+	 return err;
+      }
+   }
+   return AddressBook::SUCCESS;
+}
+
+//==============================================================================
+/*
+  DataDeviceExternal - Load this task with any devices having External 
+  connections. Not implemented for now, awaiting definition of Externals.
+*/
+//==============================================================================
+unsigned SBase::DataDeviceExternal(PMsg_p *msg)
+{
+   return 0;
```

Comment:
I recommend writing a "Not implemented" logserver message for things like this - will make them easier to find in future.